docs: guide quick start app creation using the slack cli (#2535)
slackapi/bolt-js: docs/content/getting-started.md

---
title: Quickstart with Bolt for JavaScript
sidebar_label: Quickstart
---

This quickstart guide aims to help you get a Slack app using Bolt for JavaScript up and running as soon as possible!

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

When complete, you'll have a local environment configured with a customized [app](https://github.com/slack-samples/bolt-js-getting-started-app) running that responds to a simple greeting.

:::tip[Reference for readers]

In search of the complete guide to building an app from scratch? Check out the [building an app](/building-an-app) guide.

:::

#### Prerequisites

A few tools are needed for the following steps. We recommend using the [**Slack CLI**](https://tools.slack.dev/slack-cli/) for the smoothest experience, but other options remain available.

<Tabs groupId="cli-or-terminal">
<TabItem value="cli" label="Slack CLI">

Install the latest version of the Slack CLI to get started:

- **Mac / Linux**: https://tools.slack.dev/slack-cli/guides/installing-the-slack-cli-for-mac-and-linux

> - **Windows**: https://tools.slack.dev/slack-cli/guides/installing-the-slack-cli-for-windows

Author: @mwbrooks Great suggestion! I might've been quick to paste for a larger link to target, but I like these titles 👾

Then confirm a successful installation with the following command:

```sh
$ slack version
```

An authenticated login is also required if this hasn't been done before:

```sh
$ slack login
```

</TabItem>
<TabItem value="terminal" label="Terminal">

Tooling for the terminal can also be used to follow along:

- **Git**: https://git-scm.com/downloads
- **Node.js**: https://nodejs.org/en/download

Once installed, make sure recent versions are being used:

```sh
$ git --version
$ node --version
```

</TabItem>
</Tabs>

:::info[A place to belong]

A workspace where development can happen is also needed.

We recommend using [developer sandboxes](https://docs.slack.dev/tools/developer-sandboxes) to avoid disruptions where real work gets done.

:::

> ## Initializing a project {#initializing-a-project}

Author: @mwbrooks 🧠 Aligning with the CLI commands being used makes so much sense!

This might've been carried over from earlier changes where the `init` command was used, but for this quickstart we can start from a template for development with https://github.com/slackapi/bolt-js/commit/ff6871534df2448e46803b25e34f4d3f18c9be5f 🚢 💨

```sh
$ slack create first-bolt-app --template slack-samples/bolt-js-getting-started-app
$ cd first-bolt-app
```

After a project is created you'll have a `package.json` file with project details and a `.slack` directory for application use.

A few other files exist too, but we'll visit these later.

</TabItem>
<TabItem value="terminal" label="Terminal">

A blank template can be cloned to start with project scaffolding:

```sh
$ git clone https://github.com/slack-samples/bolt-js-getting-started-app first-bolt-app
$ cd first-bolt-app
$ npm install
```

Outlines of a project are taking shape, so we can move on to creating an app!

</TabItem>
</Tabs>

## Creating an app {#creating-an-app}

Before you can start developing with Bolt, you will want to create a Slack app.

<Tabs groupId="cli-or-terminal">
<TabItem value="cli" label="Slack CLI">

The getting started app template contains a `manifest.json` file with details about an app that we will use to get started. Use the following command and select "Create a new app" to install the app to the team of choice:

```sh
$ slack run
...
[INFO]  bolt-app ⚡️ Bolt app is running!
```

> After the app has started, celebrate, then interrupt the process with `^C` to stop it for now.

Author: @mwbrooks I like this suggestion so much - following along becomes much easier in https://github.com/slackapi/bolt-js/pull/2535/commits/ff6871534df2448e46803b25e34f4d3f18c9be5f!

This section has been updated for both the **CLI** and **Terminal** sections to read:

> With the app running, you can test it out with the following steps:
>  
> 1. Open a direct message with your app or invite the bot to a public channel.
> 2. Send "hello" to the current conversation and wait for a response.
> 3. Click the attached button labelled "Click me" to post another reply.
> 
> After confirming the app responds, celebrate, then interrupt the process by pressing CTRL+C in the terminal to stop your app from running.

Because for some reason three steps is a good number for me when counting 👾✨

The created app will have some placeholder values and a small set of [scopes](https://docs.slack.dev/reference/scopes) to start, but we recommend exploring the customizations possible on app settings:

```sh
$ slack app settings
```

On these pages you're free to make changes, such as updating your app icon, or inspect various credentials:

> ![Basic Information page](/img//basic-information-page.png "Basic Information page")

Author: @mwbrooks Once more, this streamlines the ideas being shared! 📚 ✨ 

I'm not certain the current writing of this section makes more sense, but I do think introducing app settings as a place to make changes is useful to somehow keep 🤔

</TabItem>
<TabItem value="terminal" label="Browser">

> Naviage to your list of apps and [create a new Slack app](https://api.slack.com/apps/new) using the "from a manifest" option:

Author: @mwbrooks 👁️‍🗨️ ✨

```sh
$ slack version
```

An authenticated login is also required if this hasn't been done before:

```sh
$ slack login
```

</TabItem>
<TabItem value="terminal" label="Terminal">

Tooling for the terminal can also be used to follow along:

- [Git](https://git-scm.com/downloads)
- [Node.js](https://nodejs.org/en/download)

Once installed, make sure recent versions are being used:

```sh
$ git --version
$ node --version
```

</TabItem>
</Tabs>

:::info[A place to belong]

A workspace where development can happen is also needed.

We recommend using [developer sandboxes](https://docs.slack.dev/tools/developer-sandboxes) to avoid disruptions where real work gets done.

:::

## Creating a project {#creating-a-project}

With the toolchain configured, it's time to set up a new Bolt project. This contains the code that handles logic for your app.

If you don’t already have a project, let’s create a new one!

<Tabs groupId="cli-or-terminal">
<TabItem value="cli" label="Slack CLI">

A blank template can be used to start with project scaffolding:

```sh
$ slack create first-bolt-app --template slack-samples/bolt-js-getting-started-app
$ cd first-bolt-app
```

> After a project is created you'll have a `package.json` file with project details and a `.slack` directory for application use.

Reviewer: suggestion: I think we could clarify the purpose of each of these files, such as:
```suggestion
After a project is created you'll have a `package.json` file for app dependencies and a `.slack` directory for Slack CLI configuration.
```

</TabItem>
</Tabs>

## Running the app {#running-the-app}

Before you can start developing with Bolt, you will want a running Slack app.

<Tabs groupId="cli-or-terminal">
<TabItem value="cli" label="Slack CLI">

The getting started app template contains a `manifest.json` file with details about an app that we will use to get started. Use the following command and select "Create a new app" to install the app to the team of choice:

```sh
$ slack run
...
[INFO]  bolt-app ⚡️ Bolt app is running!
```

With the app running, you can test it out with the following steps:

1. Open a direct message with your app or invite the bot to a public channel.
2. Send "hello" to the current conversation and wait for a response.
3. Click the attached button labelled "Click me" to post another reply.

After confirming the app responds, celebrate, then interrupt the process by pressing `CTRL+C` in the terminal to stop your app from running.

</TabItem>
<TabItem value="terminal" label="Browser">

Navigate to your list of apps and [create a new Slack app](https://api.slack.com/apps/new) using the "from a manifest" option:

1. Select the workspace to develop your app in.

> 2. Use copied values from the `manifest.json` file to create your app.

Reviewer: suggestion: for minor clarity.
```suggestion
2. Copy and paste the `manifest.json` file to create your app.
```

3. Confirm the app features and click "Create".

You'll then land on your app's **Basic Information** page, which is an overview of your app and which contains important credentials:

![Basic Information page](/img//basic-information-page.png "Basic Information page")

To listen for events happening in Slack (such as a new posted message) without opening a port or exposing an endpoint, we will use [Socket Mode](/concepts/socket-mode). This connection requires a specific app token:

> 1. On the **Basic Information** page, scroll to the "App-Level Tokens" section and click **Generate Token and Scopes**.
> 2. Name the token "Development" or something similar and add the `connections:write` scope, then click "Generate".

Reviewer: suggestion: In the section below, I think we bold the sections a well. This makes it a little easier to read at a glance.:
```suggestion
1. On the **Basic Information** page, scroll to the **App-Level Tokens** section and click **Generate Token and Scopes**.
2. Name the token "Development" or something similar and add the `connections:write` scope, then click **Generate**.
```

1. Navigate to the **OAuth & Permissions** on the left sidebar and install your app to your workspace to generate a token.
2. After authorizing the installation, you'll return to the **OAuth & Permissions** page and find a **Bot User OAuth Access Token**:

![OAuth Tokens](/img//bot-token.png "Bot OAuth Token")

3. Copy the bot token beginning with `xoxb` from the **OAuth & Permissions page** and then store it in a new environment variable:

```sh
$ export SLACK_BOT_TOKEN=xoxb-example
```

After the app has been created, it is time to run it:

```sh
$ npm run start
...
[INFO]  bolt-app ⚡️ Bolt app is running!
```

With the app running, you can test it out with the following steps:

1. Open a direct message with your app or invite the bot to a public channel.
2. Send "hello" to the current conversation and wait for a response.
3. Click the attached button labelled "Click me" to post another reply.

After confirming the app responds, celebrate, then interrupt the process by pressing `CTRL+C` in the terminal to stop your app from running.

</TabItem>
</Tabs>

## Updating the app

> At this point, you're successfully running the getting started Bolt for JavaScript [app](https://github.com/slack-samples/bolt-js-getting-started-app)!

Reviewer: suggestion: In the previous section, we concluded with `CTRL+C` so the app isn't currently running.
```suggestion
At this point, you've successfully run the getting started Bolt for JavaScript [app](https://github.com/slack-samples/bolt-js-getting-started-app)!
```

The defaults included leave opportunities abound, so to personalize this app let's now edit the code to respond with a kind farewell.

#### Responding to a farewell

Chat is a common thing apps do and responding to various types of messages can make conversations more interesting.

Using an editor of choice, open the `app.js` file and add the following message listener after the "hello" handler:

```js
app.message('goodbye', async ({ say }) => {
  const responses = ['Adios', 'Au revoir', 'Farewell'];
  const parting = responses[Math.floor(Math.random() * responses.length)];
  await say(`${parting}!`);
});
```

> Once the file is saved, restart your app to make sure the latest changes are being used, then go back to channel and say "goodbye".

Reviewer: suggestion: Since this is a Getting Started guide, I think it's best to hold the developer's hand all the way. I'd suggest that we show the command to run your app rather than having the developer figure out how to "restart your app". This would require a `<TabItem>` for the CLI / Terminal to show the `slack run` and `npm start` commands again and explain to open a DM with their bot and send the "goodbye" text.

Author: @mwbrooks Once more a great call that found https://github.com/slackapi/bolt-js/pull/2535/commits/ef6937f48971eb48fbfb29eab0e68539e09ae392! I might be wanting to avoid repetition when writing, but that doesn't translate so well to the guide 😉

#### Customizing app settings

The created app will have some placeholder values and a small set of [scopes](https://docs.slack.dev/reference/scopes) to start, but we recommend exploring the customizations possible on app settings.

<Tabs groupId="cli-or-terminal">
<TabItem value="cli" label="Slack CLI">

> Visit app settings for your app with the following command:

Reviewer: suggestion: I'd use "Open" or "Browse to" instead of "Visit"
```suggestion
Open app settings for your app with the following command:
```

```sh
$ slack app settings
```

This will open the following page in a web browser:

![Basic Information page](/img//basic-information-page.png "Basic Information page")

</TabItem>
<TabItem value="terminal" label="Browser">

> Return back to the pages of customization for your app at a similar URL:
> 
> ```
> https://api.slack.com/apps/A0123456789
> ```

Reviewer: suggestion: I feel like new developers will paste this URL and become confused around why it's a 404 page. Instead, let's hold their hand a little more.
```suggestion
Browse to https://api.slack.com/apps and select your app from the list.

This will open the following page:

![Basic Information page](/img//basic-information-page.png "Basic Information page")
```
